msat/src/msat_solver.cpp

```diff
@@ -1264,74 +1274,93 @@ Term MsatInterpolatingSolver::get_value(const Term & t) const
+}
 
-  msat_term mA = static_pointer_cast<MsatTerm>(A)->term;
-  msat_term mB = static_pointer_cast<MsatTerm>(B)->term;
-
-  int group_A = msat_create_itp_group(env);
-  int group_B = msat_create_itp_group(env);
-
-  msat_set_itp_group(env, group_A);
-  msat_assert_formula(env, mA);
-  msat_set_itp_group(env, group_B);
-  msat_assert_formula(env, mB);
-
-  msat_result res = msat_solve(env);
+// Compute interpolation sequence with incremental solving.
+// The function tries to reuse as many previously-asserted formulas as possible.
+// To achieve this, an assertion stack is maintained, where each assertion
+// is associated with a context level (backtrack point in MathSAT) and an
+// interpolation group.
+//
+// Before SMT solving, the function scans the assertion stack and the current
+// input formulae in order, checking if they match at each position.
+// Specifically, the i-th element on the assertion stack must match the i-th
+// element in the input formulae. If they match, the assertion can be reused;
+// otherwise, the function backtracks the solver to the point where the match
+// ends and asserts the remaining formulas from that point onward.
+//
+// The folllowing invariant should hold before and after the call:
+// `#msat-backtrack-points == assertions_.size() == interp_grps_.size()`
+Result MsatInterpolatingSolver::get_sequence_interpolants(
+    const TermVec & formulae, TermVec & out_I) const
+{
+  initialize_env();
+  assert(msat_num_backtrack_points(env) == assertions_.size());
+  assert(interp_grps_.size() == assertions_.size());
 
-  if (res == MSAT_UNSAT)
+  // check which assertions can be reused
+  size_t num_reused = 0;
+  for (; num_reused < assertions_.size() && num_reused < formulae.size();
+       ++num_reused)
   {
-    msat_term itp = msat_get_interpolant(env, &group_A, 1);
-    if (MSAT_ERROR_TERM(itp))
-    {
-      throw InternalSolverException("Failed when computing interpolant.");
-    }
-    else
+    if (assertions_.at(num_reused) != formulae.at(num_reused))
     {
-      out_I = make_shared<MsatTerm>(env, itp);
-      return Result(UNSAT);
+      if (num_reused == 0)
+      {
+        // no reusable formulas, simply reset
+        msat_reset_env(env);
+      }
+      else
+      {
+        // pop formulas that cannot be reused
+        for (size_t j = num_reused; j < assertions_.size(); ++j)
+        {
+          msat_pop_backtrack_point(env);
+        }
+      }
+      break;
     }
   }
-  else if (res == MSAT_SAT)
-  {
-    return Result(SAT);
-  }
-  else
-  {
-    return Result(UNKNOWN);
-  }
-}
 
-Result MsatInterpolatingSolver::get_sequence_interpolants(
-    const TermVec & formulae, TermVec & out_I) const
-{
-  initialize_env();
-  // reset the environment -- each get_sequence_interpolants is it's own
-  // separate call
-  msat_reset_env(env);
+  // update the interpolation groups and assertions
+  assert(num_reused == msat_num_backtrack_points(env));
+  interp_grps_.resize(num_reused);
```

Comment:
I'm not entirely sure what `msat_num_backtrack_points` returns, but won't this assertion fail if we only reuse a prefix of all existing assertions? Say, if the previous query was for the sequence interpolant of `A`, `B`, `C`, and this one is just `A`, `B`. In that case, `num_reused = 2`, but what will `msat_num_backtrack_points` return?

Reply:
My previous fix was incomplete. 
2632dcb should ensure that the assertion holds.

In the above example,
`msat_num_backtrack_points` will return 3 when entering `get_sequence_interpolants`,
followed by 1 `msat_pop_backtrack_point`,
then `msat_num_backtrack_points` will return 2 at this line.